lib/driver.js

```diff
@@ -1090,7 +1215,100 @@ export class XCUITestDriver extends BaseDriver {
     if (cmd === 'getStatus') {
       return await this.getStatus();
     }
-    return await super.executeCommand(cmd, ...args);
+
+    if (cmd === 'createSession') {
+      return await super.executeCommand(cmd, ...args);
+    }
+
+    const wdaRestartEnabled = Boolean(this.opts.enableWdaRestart);
+
+    if (!wdaRestartEnabled) {
+      return await super.executeCommand(cmd, ...args);
+    }
+
+    const maxAttempts = this.opts.wdaStartupRetries || 2;
+    let attempts = 0;
+
+    while (true) {
+      try {
+        return await super.executeCommand(cmd, ...args);
+      } catch (err) {
+        this.log.debug(`Error executing command '${cmd}': ${err.message}`);
+        attempts++;
+
+        const isWdaCrash = this.isWdaCrashOrSessionInvalid(err);
+
+        if (!isWdaCrash || attempts > maxAttempts) {
+          throw err;
+        }
+
+        this._isRecovering = true;
+        this.log.warn(`WDA issue detected. Recovery attempt ${attempts}/${maxAttempts}...`);
+
+        try {
+          await this.handleWdaRecovery();
+          this.log.info('Recovery successful, retrying original command');
+        } catch (recoveryErr) {
+          this.log.error(`Recovery attempt failed: ${recoveryErr.message}`);
+          if (attempts >= maxAttempts) {
+            throw new Error(
+              `Failed to recover WDA after ${maxAttempts} attempts: ${recoveryErr.message}`,
+            );
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Checks if an error indicates a WDA crash or invalid session
+   * @private
+   * @param {Error} err - The error to check
+   * @returns {boolean} - True if the error indicates WDA crash or invalid session
+   */
+  isWdaCrashOrSessionInvalid(err) {
+    if (this.isWdaConnectionError(err)) {
+      return true;
+    }
+
+    return Boolean(
+      err.message &&
+        (err.message.includes('invalid session id') ||
+          err.message.includes('Session does not exist')),
+    );
+  }
+
+  /**
+   * Handles WDA recovery by attempting reconnection or reinitialization
+   * @private
+   * @returns {Promise<void>}
+   */
+  async handleWdaRecovery() {
+    try {
+      this.log.info('Attempting to reconnect to WDA...');
+      await this.reconnectWda();
+
+      try {
+        // Verify session is still valid
+        // This is important because in case the WDA was not killed and it was only an intermittent connection issue,
+        // and the session is still valid then this would prevent us restarting it unnecesarilly
+        await this.proxyCommand(`/session/${this.sessionId}/url`, 'GET');
```

Comment:
This endpoint call could be expensive in WebView context. Getting a timeout endpoint call could be lightweight.